### PR TITLE
Support for kernel allocation of regions of flash memory for storage

### PR DESCRIPTION
**Pull request by phil-levis**
This allocates a ROM region in the kernel image called `.storage` 
and a macro `storage_volume!` that creates variables within this
region. The region is aligned, at beginning and end, on a flash
page boundary. Together, these two allow software to allocate
on-chip flash for kernel non-volatile storage (e.g., a log).
Future pull requests will add storage abstractions that use
these mechanisms.

Note that since storage volumes are allocated as variables,
if they are not accessed/used by an abstraction, they will be
optimized away by the linker. Since you don't have any assurances
on their ordering, I think this is OK -- an unused volume being
elided won't introduce new uncertainties, since you can't be
assured they'd be at particular offsets in the first place.

### CODE REVIEW

File: kernel/src/common/utils.rs [hunk 1]
```diff
@@ -36,3 +36,29 @@ macro_rules! static_init {
...
+/// initialize their state. The linker script kernel_layout.ld makes
+/// sure that the .storage section is aligned on a 512-byte boundary
+/// and the next section is aligned as well.
+
+#[macro_export]
+macro_rules! storage_volume {
+    ($N:ident, $kB:expr) => {
+        #[link_section = ".storage"]
+        #[no_mangle]
+        pub static mut $N : [u8; $kB * 1024] = [0x00; $kB * 1024];
```
Comment (ppannuto) [inline kernel/src/common/utils.rs]: As this is in the flash region, which needs to be written as whole pages, maybe it makes more sense to not make this mutable here, it's fine to read from this buffer, but not to try and write to it. I think the eventual `sync` or `write` method should probably be the only code with mutable access to this buffer.

Comment (phil-levis) [inline kernel/src/common/utils.rs]: It's mut to make sure it's not elided by the compiler; I'm a bit wary of the degree of optimization that's applied. Am I being paranoid?

Comment (alevy) [inline kernel/src/common/utils.rs]: I don't think `mut` does that. You're not being paranoid, if a symbol isn't used it'll be elided unless the linker script `KEEP`s it.

Comment (phil-levis) [inline kernel/src/common/utils.rs]: OK, mut will go away. :)

Comment (phil-levis) [inline kernel/src/common/utils.rs]: I'm unsure whether it should use KEEP. On one hand, this means that the storage layout will change depending on whether a declared volume is referenced. E.g., if you for some reason wanted to declare a region for future use, you can't. In a C world, I can imagine wanting to be able to do this. But with Rust's safety it seems less likely -- you'd have to start trusting arbitrary memory addresses. So having volumes you don't use be pruned out (no KEEP) seems like the better option for now.

**Pull request merged**